File: src/main/java/org/tobi29/scapes/engine/opengl/vao/VAO.java
/*
 * Copyright 2012-2015 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.tobi29.scapes.engine.opengl.vao;

import org.tobi29.scapes.engine.ScapesEngine;
import org.tobi29.scapes.engine.opengl.GL;
import org.tobi29.scapes.engine.opengl.OpenGLFunction;
import org.tobi29.scapes.engine.opengl.matrix.Matrix;
import org.tobi29.scapes.engine.opengl.shader.Shader;

public abstract class VAO {
    protected final ScapesEngine engine;
    protected boolean stored, used, markAsDisposed;
    protected Runnable detach;

    protected VAO(ScapesEngine engine) {
        this.engine = engine;
    }

    public void markAsDisposed() {
        markAsDisposed = true;
    }

    @OpenGLFunction
    public boolean ensureStored(GL gl) {
        if (!stored) {
            boolean success = store(gl);
            used = true;
            return success;
        }
        used = true;
        return true;
    }

    @OpenGLFunction
    public void ensureDisposed(GL gl) {
        if (stored) {
            dispose(gl);
        }
    }

    @OpenGLFunction
    public abstract boolean render(GL gl, Shader shader);

    @OpenGLFunction
    public abstract boolean render(GL gl, Shader shader, int length);

    @OpenGLFunction
    public abstract boolean renderInstanced(GL gl, Shader shader, int count);

    @OpenGLFunction
    public abstract boolean renderInstanced(GL gl, Shader shader, int length,
            int count);

    protected abstract boolean store(GL gl);

    protected abstract void dispose(GL gl);

    protected void shader(GL gl, Shader shader) {
        Matrix matrix = gl.matrixStack().current();
        gl.activateShader(shader.programID());
        shader.updateUniforms(gl);
        int uniformLocation = shader.uniformLocation(0);
        if (uniformLocation != -1) {
            gl.setUniformMatrix4(uniformLocation, false,
                    matrix.modelView().values());
        }
        uniformLocation = shader.uniformLocation(1);
        if (uniformLocation != -1) {
            gl.setUniformMatrix4(uniformLocation, false,
                    gl.modelViewProjectionMatrix().values());
        }
        uniformLocation = shader.uniformLocation(2);
        if (uniformLocation != -1) {
            gl.setUniformMatrix3(uniformLocation, false,
                    matrix.normal().values());
        }
    }

    protected void reset() {
        assert detach != null;
        detach.run();
        detach = null;
        stored = false;
        markAsDisposed = false;
    }
}

File: gles-backend/src/commonMain/kotlin/org/tobi29/scapes/engine/backends/opengles/VAO.kt
/*
 * Copyright 2012-2018 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.tobi29.scapes.engine.backends.opengles

import org.tobi29.scapes.engine.graphics.GL
import org.tobi29.scapes.engine.graphics.GraphicsObjectSupplier
import org.tobi29.scapes.engine.graphics.Model
import org.tobi29.scapes.engine.graphics.Shader

internal abstract class VAO(protected val glh: GLESHandle) : Model {
    override val gos: GraphicsObjectSupplier get() = glh
    protected var used: Long = 0
    override var isStored = false
    protected var markDisposed = false
    override var weak = false
    protected var detach: (() -> Unit)? = null

    override fun markAsDisposed() {
        markDisposed = true
    }

    abstract override fun render(
        gl: GL,
        shader: Shader
    ): Boolean

    abstract override fun render(
        gl: GL,
        shader: Shader,
        length: Int
    ): Boolean

    abstract override fun renderInstanced(
        gl: GL,
        shader: Shader,
        count: Int
    ): Boolean

    abstract override fun renderInstanced(
        gl: GL,
        shader: Shader,
        length: Int,
        count: Int
    ): Boolean

    override fun ensureStored(gl: GL): Boolean {
        if (!isStored) {
            val success = store(gl)
            used = gl.timestamp
            return success
        }
        used = gl.timestamp
        return true
    }

    override fun ensureDisposed(gl: GL) {
        if (isStored) {
            dispose(gl)
        }
    }

    override fun isUsed(time: Long) =
        time - used < 1000000000L && !markDisposed

    protected abstract fun store(gl: GL): Boolean

    protected fun shader(
        gl: GL,
        shader: Shader
    ) {
        gl.check()
        if (shader !is ShaderGL) throw IllegalArgumentException("Shader from different implementation")
        val matrix = gl.matrixStack.current()
        shader.activate(gl)
        shader.updateUniforms(gl)
        var uniformLocation = shader.uniformLocation(0)
        if (uniformLocation != GLUniform_EMPTY) {
            glh.glUniformMatrix4fv(
                uniformLocation, false,
                matrix.modelView().values
            )
        }
        uniformLocation = shader.uniformLocation(1)
        if (uniformLocation != GLUniform_EMPTY) {
            glh.glUniformMatrix4fv(
                uniformLocation, false,
                matrix.modelViewProjection().values
            )
        }
        uniformLocation = shader.uniformLocation(2)
        if (uniformLocation != GLUniform_EMPTY) {
            glh.glUniformMatrix3fv(
                uniformLocation, false,
                matrix.normal().values
            )
        }
    }
}
